### PR TITLE
feat: add deployment solution to digital ocean app platform

### LINKED ISSUE #317: Suggestion to stop hosting server-api on k8s
Imho we should stop using K8s on Digital Ocean. The future is not how we envisioned it 2 years ago when we decided that the best will be to have all our services on K8s

more details: https://github.com/asyncapi/event-gateway/issues/161

thoughts?

### PR DESCRIPTION
**Pull request by Shurtu-gal**
**Description**

- Add workflow for deployment to digital ocean droplets
- Remove k8s helm chart files
- Add digital ocean terraform files

**Proof that it works**
- The site is hosted at [http://142.93.209.65/v1/help/validate](http://142.93.209.65/v1/help/validate)
- Successful workflow run:- https://github.com/ash17290/server-api/actions/runs/7476130464/job/20345827991

**Things to be noted by the maintainer**
- You have to add a public ssh key in https://github.com/Shurtu-gal/server-api/blob/feat/droplet/deployments/droplets/server-api.yaml#L14
- Install terraform
- Run terraform init and terraform apply
- Copy the ipv4 from output.
- Add the following secrets:
   - HOST: ipv4 address of container
   - USERNAME: asyncapi
   - PORT: 22
   - KEY: **Full Private Key for the public ssh key**

Note: this is a single-time process.

**Related issue(s)**
Fixes #317

### CODE REVIEW

**Comment by Shurtu-gal** [comment]
@derberg A small review of GitHub workflow would be appreciated.

File: terraform.tfstate
Comment (smoya) [inline terraform.tfstate:1]: tfstate should not be commited to the repository; it might contain secrets and other stuff.

**Comment by smoya** [comment]
I noticed we are following an approach of using Digital Ocean droplets directly.
Even though it might seem simple and minimalistic enough, it can be problematic in the long run. Droplets are just VM in the cloud, they not provide any service control over what is running inside, no metrics, no alarms, etc. Less important at this moment but worth to mention the fact they neither provide scale mechanisms. More [here](https://docs.digitalocean.com/products/app-platform/details/features/).
For example, managing SSL certificates is a manual thing. In our particular case, we would need to issue a new certificate (e.g. Let's Encrypt one), install it in the machine, taking care of renovation of it, etc. Something we also need to care is about application lifecycle; when do we stop a container for deploying the new one? how to rollback in case of failure? Whats the current status after an error?

I know we are migrating to anything else rather than K8s on Digital Ocean because of pricing. I completely agree that, atm, using K8s is completely far from our needs being an overengineered solution.
Luckily, Digital Ocean has the [App Platform](https://www.digitalocean.com/products/app-platform) service. In short, lets you run your containers in the cloud under a managed environment, providing application life-cycle control, monitoring (including alarms on things are wrong, such as bad deployment), domain and SSL certificate handling including automatic renewal of Let's Encrypt certificates, etc etc etc.

The pricing (in our case) should not be way more expensive than just running droplets isolated and, definitely, way more cheaper than K8s solution. The benefits are really worth.

Regarding the deployment, you just [deploy to Digital Ocean](https://docs.digitalocean.com/products/app-platform/how-to/deploy-from-container-images/) telling "hey, just deploy latest version" and the system will pick up latest container image tag and deploy it, taking care of the lifecycle from that moment on.

Additionally, we can manage the App Platform via Terraform (as you are doing with the Droplet).

cc @derberg

File: terraform.tfstate
Comment (Shurtu-gal) [reply terraform.tfstate]: It is already [there](https://github.com/asyncapi/server-api/blob/b04f43675627e69f3c95916f2664cd0fd94a8196/deployments/droplets/.gitignore#L2). This was due to running terraform here by mistake.

**Comment by Shurtu-gal** [comment]
@smoya @derberg The PR has been updated. 

Successful workflow run :- https://github.com/ash17290/server-api/actions/runs/7525186615/job/20481139953
Deployed Link :- https://server-api-ek54f.ondigitalocean.app/v1/help/validate

File: deployments/droplets/main.tf
Comment (smoya) [inline deployments/droplets/main.tf:1]: We can remove all droplets dir

File: deployments/apps/main.tf
Comment (smoya) [inline deployments/apps/main.tf:20]: Is this enough for configuring a domain + SSL management?

File: deployments/apps/main.tf
Comment (smoya) [inline deployments/apps/main.tf:51]: I don't think you need this. It is already the entry point of the Docker image.

File: deployments/apps/main.tf
Comment (smoya) [inline deployments/apps/main.tf:48]: In K8s we used /. I dont know what / returns but as far as it is a 2xx response code, is valid. Can you check? Better to not use a particular path, we will be collecting metrics soon and this healthcheck will add false hits.

File: deployments/droplets/main.tf
Comment (Shurtu-gal) [reply deployments/droplets/main.tf:1]: Okay.

File: deployments/apps/main.tf
Comment (Shurtu-gal) [reply deployments/apps/main.tf:20]: If you add the DO servers then it can be managed there itself I guess. But as we are using subdomains, I guess we will have to point CNAME to the do server.

image
image

Comment (Shurtu-gal) [reply deployments/apps/main.tf:20]: http://server-api.ashish-padhy.co/v1/help/validate
Like this

File: deployments/apps/main.tf
Comment (Shurtu-gal) [reply deployments/apps/main.tf:48]: But all other paths return `404`, due to which server deploy doesn't work. Should I add a health endpoint ?

File: deployments/apps/main.tf
Comment (Shurtu-gal) [reply deployments/apps/main.tf:20]: https://www.digitalocean.com/community/questions/app-aplication-in-laravel

Comment (smoya) [reply deployments/apps/main.tf:20]: > If you add the DO servers then it can be managed there itself I guess. But as we are using subdomains, I guess we will have to point CNAME to the do server.
> 
> image
> image
> 

api.asyncapi.com DNS are pointing already to Digital Ocean ones.
The point is, you should need a eay to tell Digital Ocean "this" is the domain for this service, please manage the SSL, etc

Comment (smoya) [reply deployments/apps/main.tf:20]: As you pointed out with https://www.digitalocean.com/community/questions/app-aplication-in-laravel, you should need to explicitly link a domain to your app in Digital Ocean.
We should replicate that into Terraform

Comment (Shurtu-gal) [reply deployments/apps/main.tf:20]: This is only applicable for domains, not sub-domain though 🤔

File: deployments/apps/main.tf
Comment (smoya) [reply deployments/apps/main.tf]: Then I dont get how k8s liveness probes were working 🤔. I guess the validate path can work for now.
Ideally we should add that health check path, but out of the scope of this issue. Feel free to open a new issue so we can follow up after.

File: deployments/apps/main.tf
Comment (Shurtu-gal) [reply deployments/apps/main.tf:20]: Yeah, http://server-api.ashish-padhy.co/v1/help/validate this one.
However, it won't work now as my digital ocean credits have expired.

Comment (smoya) [reply deployments/apps/main.tf:20]: Awesome! Thanks for testing!

**Review by derberg — COMMENTED** [review]
fyi, as @smoya is a maintainer in this repo and approves things, and obviously knows what he's doing 😄 I'm not gonna involve in review

@smoya just once merged please make it clear in https://github.com/asyncapi/server-api/issues/317 that bounty issue is completed

**Review by smoya — COMMENTED** [review]
This look Ok to me. However, we would need to manually deploy to DO servers on AsyncAPI Account.

Im on holidays, coming back next monday. I wont be able to deploy it until then and see if everything on Terraform side works as expected.

**Comment by smoya** [comment]
Digital Ocean Application has been depoyed successfully using the terraform file @Shurtu-gal created. @fmvilas added the proper `CNAME` record into `api.asyncapi.com` to `server-api-84f2u.ondigitalocean.app`, which is the internal Digital Ocean domain for server-api app. 

All seems working properly, the transition from K8s to App Platform happened smoothly. BTW, we do not have K8s on Digital Ocean anymore 🎉 cc @derberg 

Now I'm gonna proceed configuring the env vars for the GH action to work.

**Comment by smoya** [comment]
Do you believe you could update the docs of this repo mentioning how the deploy of the infra works, and how deploys of the app work as well? Can be done in a different PR, not affecting in this case  to the Bounty's timeline unless you want to do it here.

**Comment by Shurtu-gal** [comment]
Sure, I can do it here itself.

File: README.md
Comment (smoya) [inline README.md:63]:
```suggestion
2. Create a [DigitalOcean Personal Access Token](https://cloud.digitalocean.com/account/api/tokens) with `read` and `write` permissions. For more information, see [DigitalOcean's documentation](https://docs.digitalocean.com/reference/api/create-personal-access-token/).
```

File: README.md
Comment (smoya) [inline README.md:64]: This should be run being located at `./deployments/apps` directory preferably

File: README.md
Comment (smoya) [inline README.md:58]: Nice!!! 👏 👏

**Comment by smoya** [comment]
DO TOKEN secret added. Time to merge and test with a new release.

Awesome work @Shurtu-gal !!!

**Review by smoya — APPROVED** [review]

**Comment by smoya** [comment]
@all-contributors please add @Shurtu-gal for doc, infra

**Comment by aeworxet** [comment]
@asyncapi/bounty_team

**Pull request merged**
